src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+        return handle_push_none(op->a, direction) + handle_push_none(op->b, direction);
+    } else if (const Sub *op = expr.as<Sub>()) {
+        return handle_push_none(op->a, direction) - handle_push_none(op->b, flip(direction));
+    } else if (const Min *op = expr.as<Min>()) {
+        return min(handle_push_none(op->a, direction), handle_push_none(op->b, direction));
+    } else if (const Max *op = expr.as<Max>()) {
+        return max(handle_push_none(op->a, direction), handle_push_none(op->b, direction));
+    } else if (const Select *op = expr.as<Select>()) {
+        const Expr true_value = handle_push_none(op->true_value, direction);
+        const Expr false_value = handle_push_none(op->false_value, direction);
+        return select(op->condition, true_value, false_value);
+    } else if (const Mul *op = expr.as<Mul>()) {
+        if (const IntImm *constant = op->b.as<IntImm>()) {
+            if (constant->value == 1) {
+                // strip off a mul-by-one.
+                return handle_push_none(op->a, direction);
+            } else {
+                return handle_push_mul(op->a, direction, constant->value);
+            }
+        } else if (const IntImm *constant = op->a.as<IntImm>()) {
+            if (constant->value == 1) {
+                // strip off a mul-by-one.
+                return handle_push_none(op->b, direction);
+            } else {
+                return handle_push_mul(op->b, direction, constant->value);
+            }
+        }
+    } else if (const Div *op = expr.as<Div>()) {
+        if (const IntImm *constant = op->b.as<IntImm>()) {
+            if (constant->value > 0) {
+                return handle_push_div(op->a, direction, constant->value);
+            } else {
+                Expr recurse = handle_push_none(op->a, flip(direction));
+                return Div::make(std::move(recurse), op->b);
+            }
+        }
+    }
+    return expr;
+}
+
+// Taken from Simplify_Let.cpp
+class CountVarUses : public IRVisitor {
+    std::map<std::string, int> &var_uses;
+
+    void visit(const Variable *var) override {
+        var_uses[var->name]++;
+    }
+
+    void visit(const Load *op) override {
+        var_uses[op->name]++;
+        IRVisitor::visit(op);
+    }
+
+    void visit(const Store *op) override {
+        var_uses[op->name]++;
+        IRVisitor::visit(op);
+    }
+
+    using IRVisitor::visit;
+
+public:
+    CountVarUses(std::map<std::string, int> &_var_uses)
+        : var_uses(_var_uses) {
+    }
+};
+
+/*
+ * Visitor for removing terms that are completely unbounded from
+ * a min or a max.
+ */
+class StripUnboundedTerms : public IRMutator {
+public:
+    // A count of the number of unbounded vars in a given sub expression.
+    int32_t unbounded_vars = 0;
+
+private:
+    // Which direction we are approximating.
+    Direction direction;
+    // Pointer to scope for checking if a variable is bounded or not.
+    const Scope<Interval> *scope_ptr;
```

Comment:
Can this be a const ref instead? Doesn't look like it can be null

Reply:
fixed in cf591f763